### ISSUE TITLE
InstructBLIP POPE Evaluation: Identical Results for Greedy and Nucleus – Prompt Issue?

### ISSUE THREAD

**Issue by Mongodiem** [zjunlp/Deco#13]
<h3>Question</h3>
<p>Hi authors, thanks for the great work!</p>
<p>I'm currently evaluating InstructBLIP with the DeCo decoding strategy on the POPE benchmark (popular, adversarial, random subsets), and I noticed that the <strong>results for greedy and nucleus decoding are identical</strong> across all subsets. I suspect it might be due to the <strong>prompt format or decoding parameters</strong>, and would appreciate your clarification.</p>
<hr>
<h3>Prompt used:</h3>
<pre><code class="language-text">&lt;image&gt;
Question: {question}
Answer (one word):
</code></pre>
<hr>
<h3>Inference code (simplified):</h3>
<pre><code class="language-python">prompt = f"&lt;image&gt;\nQuestion: {question}\nAnswer (one word):"
inputs = processor(images=image, text=prompt, return_tensors="pt").to(device)

with torch.inference_mode():
    outputs = model.generate(
        **inputs,
        do_sample=args.temperature &gt; 0,
        temperature=max(args.temperature, 1.0),
        top_p=args.top_p,
        num_beams=args.num_beams,
        max_new_tokens=args.max_new_tokens,
        use_deco=True,
        alpha=args.alpha,
        threshold_top_p=args.threshold_top_p,
        threshold_top_k=args.threshold_top_k,
        early_exit_layers=list(range(args.start_layer, args.end_layer)),
        return_dict_in_generate=True,
        output_hidden_states=True,
    )
</code></pre>
<hr>
<h3> Evaluation Results (F1, Precision, Recall)</h3>

Strategy | Subset | Temp | F1 Score | Precision | Recall | Yes Ratio
-- | -- | -- | -- | -- | -- | --
greedy | random | -1 | 0.8875 | 0.9079 | 0.8680 | 0.4923
greedy | popular | -1 | 0.8302 | 0.7978 | 0.8653 | 0.5423
greedy | adversarial | -1 | 0.8133 | 0.7671 | 0.8653 | 0.5640
nucleus | random | 1 | 0.8875 | 0.9079 | 0.8680 | 0.4923
nucleus | popular | 1 | 0.8302 | 0.7978 | 0.8653 | 0.5423
nucleus | adversarial | 1 | 0.8133 | 0.7671 | 0.8653 | 0.5640

<hr>
<h3> Question Recap</h3>
<blockquote>
<p>Is this because the prompt restricts outputs too tightly (e.g., enforcing “yes” or “no”)?<br>
Or is it due to generation settings like <code inline="">num_beams=1</code>, <code inline="">temperature=1</code>, <code inline="">top_p=0.9</code>?</p>
</blockquote>
<p>Any guidance or suggestions to <strong>diversify decoding outcomes</strong> would be appreciated!</p>
<p>Thanks 🙏</p>
<hr>

</body>
</html>

**Comment by sunnychenxiwang** [comment]
In the process of using Deco, a truncation method like top-p is first applied to truncate the vocabulary and obtain a candidate tokens, and the logits of the tokens in this candidate are reordered. If the chosen threshold_top_p is too small, there will always be only one or two tokens in the candidate list. Therefore, this issue might be caused by your prompt. You could try increasing the threshold_top_p to something like 0.95, or consider adjusting the prompt.

**Comment by Mongodiem** [comment]
Thank you for the explanation!

I tried increasing threshold_top_p as suggested, and also experimented with a few alternative prompts (e.g., removing "Answer (one word):", or using more open-ended formats), but unfortunately the results got worse—either too many hallucinations or inconsistent answers.

Would it be possible for you to kindly share the exact prompt you used in your POPE evaluation (if available)?
It would really help me align better with the intended setup and ensure a fair comparison.

Much appreciated!

**Comment by sunnychenxiwang** [comment]
I am using the original prompts from the PoPE dataset, such as: "Is there a snowboard in the image?"

**Comment by Mongodiem** [comment]
<html>
<body>

<hr>
<p>Thanks again for your help!</p>
<p>I changed the prompt from:</p>
<pre><code>&lt;image&gt;
Question: {question}
Answer (one word):
</code></pre>
<p>to a looser version:</p>
<pre><code>Question: {question}.
</code></pre>
<p>Then I ran POPE (popular) evaluation with the following results:</p>

Strategy | Beams | Temp | F1 | Yes Ratio
-- | -- | -- | -- | --
Greedy | 1 | -1 | 0.8045 | 0.6287
Nucleus | 1 | 1 | 0.8033 | 0.6287
Beam | 5 | -1 | 0.8045 | 0.6287


<ul>
<li>
<p><strong>Greedy and Beam gave identical results</strong></p>
</li>
<li>
<p><strong>Nucleus only changed slightly</strong></p>
</li>
<li>
<p>Yes ratio increased noticeably after prompt change</p>
</li>
</ul>
<p>Would you mind sharing the exact prompt used in your POPE evaluation?</p>
<p>Thanks a lot! 🙏</p>
<hr>

</body>
</html>

**Comment by sunnychenxiwang** [comment]
As shown in the provided code pope_llava.py, we did not make any modifications to the prompt. If you are using the Hugging Face version of the model for inference, here is a code example.

```python
prompt = "Is there a snowboard in the image?"
inputs = processor(images=image, text=prompt, return_tensors="pt").to(device)

outputs = model.generate(
        **inputs,
        do_sample=False,
        num_beams=5,
        max_length=256,
        min_length=1,
        top_p=0.9,
        temperature=1,
)
```

**Comment by zxlzr** [comment]
hi, do you have any further questions?

**Comment by Mongodiem** [comment]
Thank you so much for your kind reply! I followed your prompt exactly, and I’m thrilled to share that I’ve successfully reproduced about half of the results — they align remarkably well with those in your paper! This is incredibly exciting for me. I don’t have any further questions at the moment. Huge thanks again!
